subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <net/wifi_mgmt_ext.h>
+#include <net/wifi_credentials.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/net_core.h>
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/wifi_mgmt.h>
+
+/* foward declaration of supp_api stuff */
+struct add_network_resp {
+	int network_id;
+};
+int call_wpa_cli(const char *fmt, ...);
+int wpa_cli_api_add_network(struct add_network_resp *resp);
+
+/* Ensure 'strnlen' is available even with -std=c99. */
+size_t strnlen(const char *buf, size_t bufsz);
+
+LOG_MODULE_REGISTER(wifi_mgmt_ext, CONFIG_WIFI_MGMT_EXT_LOG_LEVEL);
+
+static inline const char *wpa_supp_security_txt(enum wifi_security_type security)
+{
+	switch (security) {
+	case WIFI_SECURITY_TYPE_NONE:
+		return "OPEN";
+	case WIFI_SECURITY_TYPE_PSK:
+		return "WPA-PSK";
+	case WIFI_SECURITY_TYPE_PSK_SHA256:
+		return "WPA-PSK-SHA256";
+	case WIFI_SECURITY_TYPE_SAE:
+		return "SAE";
+	case WIFI_SECURITY_TYPE_UNKNOWN:
+	default:
+		return "UNKNOWN";
+	}
+}
+
+static int add_network_from_credentials_struct_personal(struct wifi_credentials_personal *creds)
+{
+	int ret = 0;
+	struct add_network_resp resp = {0};
+
+	ret = wpa_cli_api_add_network(&resp);
+	if (ret) {
+		LOG_ERR("Encoding error: %d returned at %s:%d", ret, __FILE__, __LINE__);
+		return ret;
+	}
+
+	call_wpa_cli("set_network %d ssid \"%s\"", resp.network_id, creds->header.ssid);
+	call_wpa_cli("set_network %d key_mgmt %s", resp.network_id,
+		     wpa_supp_security_txt(creds->header.type));
+
+	if (creds->header.type == WIFI_SECURITY_TYPE_PSK ||
+	    creds->header.type == WIFI_SECURITY_TYPE_PSK_SHA256) {
+		call_wpa_cli("set_network %d psk \"%s\"", resp.network_id, creds->password);
+	}
+
+	if (creds->header.type == WIFI_SECURITY_TYPE_SAE) {
+		call_wpa_cli("set_network %d sae_password \"%s\"",
+			     resp.network_id, creds->password);
+	}
+
+	if (creds->header.flags & WIFI_CREDENTIALS_FLAG_BSSID) {
+		call_wpa_cli("set_network %d bssid %02hhx:%02hhx:%02hhx:%02hhx:%02hhx:%02hhx",
+			resp.network_id,
+			creds->header.bssid[0], creds->header.bssid[1],
+			creds->header.bssid[2], creds->header.bssid[3],
+			creds->header.bssid[4], creds->header.bssid[5]);
+	}
+
+	if (creds->header.flags & WIFI_CREDENTIALS_FLAG_FAVORITE) {
+		call_wpa_cli("set_network %d priority 1", resp.network_id);
+	}
+
+	if (creds->header.flags & WIFI_CREDENTIALS_FLAG_2_4GHz) {
+		call_wpa_cli("set_network %d freq_list "
+			     "2412 2417 2422 2427 2432 2437 2442 "
+			     "2447 2452 2457 2462 2467 2472 2484", resp.network_id);
+	}
+
+	if (creds->header.flags & WIFI_CREDENTIALS_FLAG_5GHz) {
+		call_wpa_cli("set_network %d freq_list "
+			     "5180 5200 5220 5240 5260 5280 5300 "
+			     "5310 5320 5500 5520 5540 5560 5580 "
+			     "5600 5620 5640 5660 5680 5700 5745 "
+			     "5765 5785 5805 5825", resp.network_id);
```

Comment:
Is there any way to get these frequencies from wifi_mgmt, or from hostap, rather than have them hard-coded here?

Reply:
Neither of them seem to expose them afaict. This is just needed because there is no proper interface for band selection in wpa_supplicant.

Reply:
I looked a bit deeper into `freq_list`. Apparently, if the underlying `freq_list` is set to NULL, this disables frequency checking altogether (IE, it accepts all frequencies possible). (see `freq_allowed`)

I haven't checked all of the code-base to make sure this is actually the case, but it seems worth trying. This also applies only to scanning AFAICT, but I think that's what you care about here. Correct me if I'm wrong on that.

The tricky bit is figuring out what to supply `call_wpa_cli` to actually get the `freq_list` to be NULL.

AFAICT, the best way to do this is:
`call_wpa_cli("set_network %d freq_list 0", resp.network_id);`

But please actually verify that this works!!!
`wpa_config_process_freq_list` is where the special case that the first frequency is `0` is detected.

If this doesn't work, alternatives include:
-  my suggestion in https://github.com/nrfconnect/sdk-nrf/pull/9261#discussion_r1080712094
- scanning `ieee80211_chan_to_freq` or `ieee80211_chan_to_freq_global` to generate a list

Personally, disabling the list altogether makes the most sense to me

Reply:
So for the scanning part: When requesting connection, wpa_supp will do an automated scan. When adding a network (with the ssid object), you can only set a specific channel or all channels. As a workaround, I set the global scan parameters. I'll check again how they are reset and whether we have to do that manually.

Reply:
Okay, I now know that the scanning parameters are not persisted after a disconnect request.

Reply:
As @glarsennordic says better to leave this to the underlying driver/chipset, as these vary across countries. The user can optionally pass a channel of his choice, see the [latest sdk-hostap to](https://github.com/nrfconnect/sdk-hostap/pull/57/commits/74936283966aeb742049112e692b39d9c4a1c6c8)

Reply:
You don't need to set `freq_list`, instead just leave it to the underlying chipset to figure out the channels. Only when user specific a channel in the connect command you need to configure the scan_freq to limit to a single frequency.

Reply:
How do I specify a band then? Is there an interface for this? I couldn't find one.